src/pip/_internal/download.py

```diff
@@ -933,6 +933,29 @@ def unpack_http_url(
             os.unlink(from_path)
 
 
+def _copy_source_tree(source, target):
```

Comment:
I think it would be good to add a comment saying that this is essentially a patched version of `shutil.copytree()` to fix what's arguably a bug in that function (per [this comment](https://github.com/pypa/pip/issues/5306#issuecomment-383355379)), and then also link to the line in the CPython code base mentioned in that comment.

I would also add a docstring and in the docstring say that the directory is still copied (except for paths that error out) even if errors occur on some paths.

Reply:
Even better would be to create an issue in Python’s bug tracker about this and instead link to that (the issue can include a link to the line of code).

Reply:
The comment about `shutil.copytree()` is referring to the fact that socket files do not raise `SpecialFileError`, but that is independent of how we treat the errors that get raised. I think there's a better way to structure this:

1. Wrap `copytree` and convert instances of `Error` that refer to socket files to `SpecialFileError` - this probably belongs in `pip._internal.utils.filesystem` and should have a comment/description like you're describing
2. Invoke our wrapper and ignore any instances of `SpecialFileError` in the list of exceptions, raising any other errors that occur - this would be `_copy_source_tree`

Reply:
Created [bpo-37700](https://bugs.python.org/issue37700) and semi-related [bpo-37701](https://bugs.python.org/issue37701).

Reply:
Currently `copytree()` (via `copy2()` via `copyfile()`) raises a `SpecialFileError` if either the source or the destination meets the criteria. Is it OK to assume that because of our specific usage that the destination file will never be the actual cause?

Reply:
I think so. And you could document the assumption in the docstring so people know the limitation. Also, it might be good to log a warning when handling the error so these won’t pass silently in either case.

Reply:
Done and done [here](https://github.com/pypa/pip/pull/6802/files#diff-34481ab2fefef95e8820a46e86004ae7R977).